compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -979,6 +979,17 @@ impl<'a> Parser<'a> {
                 // we are using noexpect here because we don't expect a `.` directly after a `return`
                 // which could be suggested otherwise
                 self.eat_noexpect(&token::Dot)
+            } else if self.token.kind == TokenKind::RArrow
+                && self.prev_token.span.overlaps_or_adjacent(self.token.span)
```

Comment:
very likely this needs `&& self.may_recover()` to prevent stable code like the following from regressing:

```rs
macro_rules! ck {
    ($e:expr) => { compile_error!(""); };
    ($l:ident->$r:ident) => {};
}
ck!(a->b)
```

Reply:
cc #103534

Reply:
Good idea, thanks!